Drift/Private/DriftBase.cpp

```diff
@@ -4639,6 +4644,59 @@ void FDriftBase::GetUserIdentitiesByName(const FString& name, const FDriftGetUse
     GetUserIdentitiesByNames(namesArray, delegate);
 }
 
+void FDriftBase::GetFriendRichPresence(int32 FriendId, const FDriftGetFriendRichPresenceDelegate& Delegate)
+{
+    if (state_ != DriftSessionState::Connected)
+    {
+        DRIFT_LOG(Base, Warning, TEXT("Attempting to get rich presence without being connected"));
+        (void) Delegate.ExecuteIfBound(false, {});
+        return;
+    }
+
+    FString Url = driftEndpoints.template_richpresence.Replace(TEXT("{player_id}"), *FString::FromInt(FriendId));
+
+    const auto Request = GetGameRequestManager()->Get(Url);
+    Request->OnResponse.BindLambda([this, Delegate, FriendId](ResponseContext& Context, JsonDocument& Doc)
+    {
+        FRichPresenceResult Result;
+        if (!JsonArchive::LoadObject(Doc, Result))
+        {
+            Context.error = TEXT("Failed to parse user identities response");
+            return;
+        }
+
+        RichPresenceCache.Add(FriendId, Result);
+        (void) Delegate.ExecuteIfBound(true, Result);
+    });
+
+    Request->OnError.BindLambda([this, Delegate, Url](ResponseContext& Context)
+    {
+        FString Error;
+        Context.errorHandled = GetResponseError(Context, Error);
+        DRIFT_LOG(Base, Error, TEXT("Failed to get richpresence from url: %s. Error: '%s'"), *Url, *Error);
+    });
+}
+
+void FDriftBase::CacheFriendsRichPresence(const FDriftGetFriendsRichPresenceDelegate& Delegate)
+{
+    TArray<FDriftFriend> DriftFriends;
+    GetFriendsList(DriftFriends);
+    const int32 NumFriends = DriftFriends.Num();
+    int32 Results = 0;
+
+    for (auto& DriftFriend : DriftFriends)
+    {
+        GetFriendRichPresence(DriftFriend.playerID, FDriftGetFriendRichPresenceDelegate::CreateLambda([&Results, NumFriends, Delegate](bool Success, const FRichPresenceResult& Result)
+        {
+            Results += 1;
+            if (Results == NumFriends)
```

Comment:
I might be missing something here, but is there a reason why the amount of friends being equal to the amount from `DriftFriends` is necessary for the caching process to be considered a success?

Reply:
It's a way of collating lots of small requests. Each request to drift-base is getting rich presence information for 1 friend.

So for example if you have 10 friends, you only want to trigger the outer delegate once all 10 rich presences have been requested. There isn't another good way to "await" a series of network calls like this. 

If we would be using `TFutures` or `Coroutines` or something than you could do something like a `PromiseAll(...)` implementation, but we're not.

Reply:
Alright nevermind then